presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DereferencePushDownProject.java

```diff
@@ -0,0 +1,83 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.sql.parser.SqlParser;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule.Context;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.DereferenceExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.sql.planner.SymbolsExtractor.extractAll;
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class DereferencePushDownProject
+        extends DereferencePushDownRule<ProjectNode>
+{
+    public DereferencePushDownProject(Metadata metadata, SqlParser sqlParser)
+    {
+        super(metadata, sqlParser, project());
+    }
+
+    @Override
+    protected PlanNode dereferencePushDown(Context context, ProjectNode projectNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments)
+    {
+        List<Symbol> outputSymbols = projectNode.getOutputSymbols();
+        Map<Expression, Symbol> pushdownExpressions = expressions.entrySet().stream()
+                .filter(entry -> !isCastRowType(entry.getKey()))
+                .filter(entry -> outputSymbols.contains(getOnlyElement(extractAll(entry.getKey()))))
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        if (pushdownExpressions.isEmpty()) {
+            return projectNode;
+        }
+
+        Map<Symbol, Expression> pushdownDereferences = pushdownExpressions.entrySet().stream().collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+        ImmutableMap.Builder<Symbol, Symbol> symbolsBuilder = ImmutableMap.builder();
+        for (Map.Entry<Expression, Symbol> entry : pushdownExpressions.entrySet()) {
+            symbolsBuilder.put(getOnlyElement(extractAll(entry.getKey())), entry.getValue());
+        }
+        Map<Symbol, Symbol> symbolsMap = symbolsBuilder.build();
+
+        Assignments.Builder sourceBuilder = Assignments.builder();
+        for (Map.Entry<Symbol, Expression> entry : projectNode.getAssignments().entrySet()) {
+            if (symbolsMap.containsKey(entry.getKey())) {
+                Symbol targetSymbol = symbolsMap.get(entry.getKey());
+                DereferenceExpression targetDereference = (DereferenceExpression) pushdownDereferences.get(targetSymbol);
+                if (entry.getValue() instanceof DereferenceExpression) {
+                    sourceBuilder.put(targetSymbol, targetDereference);
```

Comment:
Why is this correct? I put a breakpoint here and ran TestDereferencePushDown, but it wasn't hit. Could you add test coverage for this code path?